drake/solvers/Optimization.h

```diff
@@ -0,0 +1,323 @@
+     */
+//    void addConstraint(const std::shared_ptr<LinearConstraint>& con, const VariableList& vars) {}
+
+    /** addLinearEqualityConstraint
+     * @brief adds linear equality constraints to the program for all (currently existing) variables
+     */
+    template <typename DerivedA,typename DerivedB>
+    std::shared_ptr<LinearEqualityConstraint> addLinearEqualityConstraint(const Eigen::MatrixBase<DerivedA>& Aeq,const Eigen::MatrixBase<DerivedB>& beq) {
+      assert(Aeq.cols() == num_vars);
+      VariableList vars;
+      for (auto const& v : variables) {
+        vars.push_back(DecisionVariableView(v));
+      }
+      return addLinearEqualityConstraint(Aeq,beq,vars);
+    }
+
+  /** addLinearEqualityConstraint
+   * @brief adds linear equality constraints referencing potentially a subset of the decision variables.
+   * Example: to add and equality constraint which only depends on two of the elements of x, you could use
+   *   auto x = prog.addContinuousDecisionVariable(6,"myvar");
+   *   prog.addLinearEqualityConstraint(Aeq,beq,{x.row(2),x.row(5)});
+   * where Aeq has exactly two columns.
+   */
+    template <typename DerivedA,typename DerivedB>
+    std::shared_ptr<LinearEqualityConstraint> addLinearEqualityConstraint(const Eigen::MatrixBase<DerivedA>& Aeq,const Eigen::MatrixBase<DerivedB>& beq, const VariableList& vars) {
+      problem_type.reset(problem_type->addLinearEqualityConstraint());
+
+      auto constraint = std::make_shared<LinearEqualityConstraint>(Aeq,beq,vars);
+      linear_equality_constraints.push_back(constraint);
+      return constraint;
+    }
+
+    // template <typename FunctionType>
+    // void addCost(std::function..);
+    // void addLinearCost(const Eigen::MatrixBase<Derived>& c,const vector<const DecisionVariable&>& vars)
+    // void addQuadraticCost ...
+
+//    void addConstraint(const LinearConstraint& constraint, const std::vector<const DecisionVariable&>& vars);
+//  void addConstraint(const BoundingBoxConstraint& constraint, const std::vector<const DecisionVariable&>& vars);
+
+//    template <typename DerivedLB,typename DerivedUB>
+//    void addBoundingBoxConstraint(const Eigen::MatrixBase<DerivedLB>& lower_bound, const Eigen::MatrixBase<DerivedUB>& upper_bound, const DecisionVariable& var) {  }
+
+
+    bool solve() { return problem_type->solve(*this); }; // todo: add argument for options
+
+//    template <typename Derived>
+//    bool solve(const Eigen::MatrixBase<Derived>& x0);
+
+//    getObjectiveValue();
+//    getExitFlag();
+//    getInfeasibleConstraintNames();
+
+    void printSolution() {  // todo: overload the ostream operator instead?
+      for (const auto& v : variables) {
+        std::cout << v.name << " = " << v.value.transpose() << std::endl;
+      }
+    }
+
+
+  private:
+    template <typename Derived>
+    void setSolution(const Eigen::MatrixBase<Derived>& x) {
```

Comment:
change the function name to 'setDecisionVariableValue`? One could use this function to set the initial guess of each `DecisionVariable`, that is not to set the solution.